packages/devtools_app/lib/src/debugger/evaluate.dart

```diff
@@ -274,55 +284,232 @@ class _ExpressionEvalFieldState extends State<ExpressionEvalField>
   }
 }
 
+// If Dart had union types, ref would be type ClassRef | FuncRef | LibraryRef
+Future<LibraryRef> findOwnerLibrary(
+    Object ref, DebuggerController controller) async {
+  if (ref is LibraryRef) {
+    return ref;
+  }
+  if (ref is ClassRef) {
+    final clazz = await classFor(ref, controller);
+    return clazz?.library;
+  }
+  if (ref is FuncRef) {
+    return findOwnerLibrary(ref.owner, controller);
+  }
+  return null;
+}
+
 Future<List<String>> autoCompleteResultsFor(
   EditingParts parts,
   DebuggerController controller,
 ) async {
   final result = <String>{};
   if (!parts.isField) {
-    result.addAll(
-        controller.variables.value.map((variable) => variable.boundVar.name));
+    final variables = controller.variables.value;
+    result.addAll(variables.map((variable) => variable.boundVar.name));
+
+    final thisVariable = variables.firstWhere(
+        (variable) => variable.boundVar.name == 'this',
+        orElse: () => null);
+    if (thisVariable != null) {
+      final thisValue = thisVariable.boundVar.value;
+      if (thisValue is InstanceRef) {
+        await _addAllInstanceMembersToAutocompleteList(
+          result,
+          thisValue,
+          controller,
+        );
+      }
+    }
+    final frame = controller.frameForEval;
+    if (frame != null) {
+      final function = frame.function;
+      if (function != null) {
+        final libraryRef = await findOwnerLibrary(function, controller);
```

Comment:
Why must we go through the function? I thought a frame has a corresponding library reference.

Reply:
I agree this is annoying but I fail to see where the library is on the frame. You can go to the source file directly but that is not the same thing as the library.
@bkonyi  am I missing something?
https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#frame

Reply:
As of `3.46` you can get access to the `@Library` of a `@Function` by doing something like `function.owner as @Library` (for top level functions) or `(function.owner as @Class).library`. That functionality just landed this week though, so most applications in the wild will require requesting the full objects to eventually get to the `@Library` in either the `Class` or `Function` object.

Reply:
Thanks! I'm now taking advantage of that where possible.